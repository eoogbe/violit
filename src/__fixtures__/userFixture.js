/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const username = 'fred-astaire';

export const password = 'password123';

export const credentials = {
  username,
};

export const authUrl = 'http://localhost:4000/api/auth';

export const usersUrl = 'http://localhost:4000/api/users';

export const userUrl = 'http://localhost:4000/api/users/fred-astaire';

export const authApi = [
  {
    href: authUrl,
    data: [
      { name: 'username', value: username },
    ],
  }
];

export const userApi = [
  {
    href: userUrl,
    data: [
      { name: 'username', value: username },
    ],
  }
];

export const credentialsFetch = {
  url: authUrl,
  method: 'GET',
  status: 200,
  body: {
    items: userApi,
  },
};

export const loginFetch = {
  url: authUrl,
  method: 'POST',
  status: 201,
  body: {
    items: userApi,
  },
};

export const logoutFetch = {
  url: authUrl,
  method: 'DELETE',
  status: 200,
};

export const signupFetch = {
  url: usersUrl,
  method: 'POST',
  status: 201,
  body: {
    items: userApi,
  },
};

export const hasUserFetch = {
  url: userUrl,
  method: 'HEAD',
  status: 200,
};

export const doesNotHaveUserFetch = {
  url: userUrl,
  method: 'HEAD',
  status: 404,
};
