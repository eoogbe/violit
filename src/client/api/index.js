/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
