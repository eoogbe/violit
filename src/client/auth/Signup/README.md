```jsx
import { MemoryRouter } from 'react-router-dom';

<StoreProvider>
  <MemoryRouter initialEntries={['/signup']}>
    <Signup createAsyncFunction={() => {}} />
  </MemoryRouter>
</StoreProvider>
```
