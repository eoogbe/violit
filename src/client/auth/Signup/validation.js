/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/** Validates that the username is correct. */
export const validateUsername =
  async ({ username }, { checkHasUserAsyncFunc }) => {
    if (!username) {
      return 'Required.';
    }

    if (!/^[\w-]+$/.test(username)) {
      return 'Can only have letters, numbers, \'-\', or \'_\'.';
    }

    if (!/^[A-Z]/i.test(username)) {
      return 'Must start with a letter.';
    }

    try {
      const { hasUser } = await checkHasUserAsyncFunc.asyncFunction({
        username,
      });
      if (hasUser) {
        return 'Already taken.';
      }
    } catch (e) {
      // Pass through. Handle validation on the server.
    }

    return null;
  };

/** Validates that the password is correct. */
export const validatePassword = ({ password }) => {
  if (!password) {
    return 'Required.';
  }

  if (password.length < 8) {
    return 'Must be at least 8 characters.';
  }

  return null;
};

/** Validates that the password confirmation is  correct. */
export const validatePasswordConfirmation =
  ({ password, passwordConfirmation }) => {
    if (!passwordConfirmation) {
      return 'Required.';
    }

    if (passwordConfirmation !== password) {
      return 'Must match password.';
    }

    return null;
  };
