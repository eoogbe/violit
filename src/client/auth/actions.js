/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import createActions from 'client/utils/createActions';

export default createActions('auth', [
  'LOAD_TRIGGERED',
  ['LOAD_SUCCEEDED', 'user'],
  'LOAD_FAILED',
  ['LOGIN_TRIGGERED', 'credentials'],
  ['LOGIN_SUCCEEDED', 'user'],
  'LOGIN_FAILED',
  ['CREDENTIALS_INVALID', 'error'],
  'LOGOUT_TRIGGERED',
  ['SIGNUP_TRIGGERED', 'user'],
  ['SIGNUP_SUCCEEDED', 'user'],
  ['SIGNUP_FAILED'],
  ['CHECK_HAS_USER_TRIGGERED', 'username'],
  ['CHECK_HAS_USER_SUCCEEDED', 'hasUser'],
  'CHECK_HAS_USER_FAILED',
]);
