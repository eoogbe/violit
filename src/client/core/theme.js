/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default {
  primaryColor: '#7F00FF',

  gray200: '#EEEEEE',
  gray300: '#E0E0E0',
  gray500: '#9E9E9E',
  gray700: '#616161',

  red200: '#EF9A9A',
  red500: '#F44336',

  lightBlue700: '#0288D1',

  black: '#000000',
  white: '#FFFFFF',

  spacingSm: '0.25em',
  spacingMd: '0.5em',
  spacingLg: '1em',

  borderRadius: '2px',
};
