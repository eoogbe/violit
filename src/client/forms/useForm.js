/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import camelCase from 'camel-case';
import * as R from 'ramda';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

/** Hook that adds state for a form with controlled components. */
export default (initialValues, validate, options) => {
  const { commitOnChange } = options || {};

  const prevValues = useRef(initialValues);
  const preventDoubleSubmit = useRef(false);
  const [values, setValues] = useState(initialValues);
  const [errors, setErrors] = useState({});
  const [dirtied, setDirtied] = useState(R.map(R.F, initialValues));
  const [submitting, setSubmitting] = useState(false);

  const unmounted = useRef(false);
  useEffect(() => () => {
    unmounted.current = true;
  }, []);

  const isDirty = useMemo(() => {
    return Object.values(dirtied).includes(true);
  }, [dirtied]);

  const hasError = useMemo(() => {
    return Object.values(errors).some((error) => !!error);
  }, [errors]);

  const validateFields = useCallback(async () => {
    if (!validate) {
      return {};
    }

    const errors = await validate(values) || {};
    if (!unmounted.current) {
      setErrors(errors);
    }

    return errors;
  }, [validate, values]);

  useEffect(() => {
    if (values === prevValues.current) {
      return;
    }

    validateFields();
    prevValues.current = values;
  }, [values, validateFields]);

  const reset = useCallback(() => {
    if (unmounted.current) {
      return;
    }

    prevValues.current = initialValues;
    preventDoubleSubmit.current = false;
    setValues(initialValues);
    setErrors({});
    setDirtied(R.map(R.F, initialValues));
    setSubmitting(false);
  }, [initialValues]);

  const handleInput = useCallback(({ target }) => {
    const { name, value } = target;
    setValues(R.assoc(camelCase(name), value));
  }, []);

  const handleCommit = useCallback(({ target }) => {
    const { value } = target;
    const name = camelCase(target.name);
    setValues(R.assoc(name, value));
    setDirtied(R.assoc(name, true));
  }, []);

  const handleChange = useCallback((event) => {
    handleInput(event);
    if (commitOnChange) {
      handleCommit(event);
    }
  }, [handleInput, handleCommit, commitOnChange]);

  const handleSubmit = useCallback((onSubmit) => async (event) => {
    event.preventDefault();

    if (preventDoubleSubmit.current) {
      return;
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = true;
      setSubmitting(true);
    }

    const fieldErrors = await validateFields();
    if (Object.values(fieldErrors).every((error) => !error)) {
      const submitErrors = await onSubmit(values) || {};
      if (!unmounted.current) {
        setErrors(submitErrors);
      }
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = false;
      setSubmitting(false);
    }
  }, [validateFields, values]);

  const eventHandlers = {
    handleChange,
    handleBlur: handleCommit,
    handleSubmit,
  };

  return [
    values,
    errors,
    dirtied,
    eventHandlers,
    { submitting, isDirty, hasError, reset, validateFields },
  ];
};
