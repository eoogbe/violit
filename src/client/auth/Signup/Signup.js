/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import * as R from 'ramda';
import { useHistory, useLocation } from 'react-router-dom';
import styled from 'styled-components';
import { Button, useTitle } from 'client/core';
import { FormError, Input, useForm } from 'client/forms';
import actions from '../actions';
import {
  validateUsername,
  validatePassword,
  validatePasswordConfirmation,
} from './validation';

const StyledSignup = styled.form`
  font-size: 1rem;
  padding: ${({ theme }) => theme.spacingMd};
  display: flex;
  flex-direction: column;
`;

const Heading = styled.h2`
  font-size: 1.75rem;
  margin: 0;
  margin-bottom: 0.288em;
`;

const FieldList = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
`;

FieldList.displayName = 'FieldList';

const SubmitButton = styled(Button)`
  align-self: flex-end;
`;

/** The screen for registering a new user. */
const Signup = ({ createAsyncFunction }) => {
  const history = useHistory();
  const location = useLocation();

  useTitle('Signup');

  const checkHasUserAsyncFunc = createAsyncFunction({
    start: actions.checkHasUserTriggered.type,
    resolve: actions.checkHasUserSucceeded.type,
    reject: actions.checkHasUserFailed.type,
  });
  const [
    values,
    errors,
    dirtied,
    { handleChange, handleBlur, handleSubmit },
    { submitting, isDirty, hasError },
  ] = useForm({
    username: '',
    password: '',
    passwordConfirmation: '',
  }, async (values) => ({
    username: await validateUsername(values, { checkHasUserAsyncFunc }),
    password: await validatePassword(values),
    passwordConfirmation: await validatePasswordConfirmation(values),
  }));
  const disabled = !isDirty || hasError || submitting;

  const signupAsyncFunc = createAsyncFunction({
    start: actions.signupTriggered.type,
    resolve: actions.signupSucceeded.type,
    reject: actions.credentialsInvalid.type,
  });
  const onSubmit = async (values) => {
    const user = R.pick(['username', 'password'], values);
    try  {
      await signupAsyncFunc.asyncFunction({ user });
    } catch (err) {
      return {
        form: err.error,
      };
    }

    const from = R.pathOr('/', ['state', 'from'], location);
    history.replace(from);
  };

  return (
    <StyledSignup onSubmit={handleSubmit(onSubmit)}>
      <Heading>Signup</Heading>
      {errors.form && <FormError>{errors.form}</FormError>}
      <FieldList>
        <li>
          <label htmlFor="username">Username</label>
          <Input
            id="username"
            type="text"
            name="username"
            aria-required="true"
            value={values.username}
            onChange={handleChange}
            onBlur={handleBlur}
            autoFocus
          />
          {dirtied.username && errors.username &&
            <FormError aria-describedby="username">
              {errors.username}
            </FormError>}
        </li>
        <li>
          <label htmlFor="password">Password</label>
          <Input
           id="password"
           type="password"
           name="password"
           aria-required="true"
           value={values.password}
           onChange={handleChange}
           onBlur={handleBlur}
          />
          {dirtied.password && errors.password &&
            <FormError aria-describedby="password">
              {errors.password}
            </FormError>}
        </li>
        <li>
          <label htmlFor="password-confirmation">Password confirmation</label>
          <Input
           id="password-confirmation"
           type="password"
           name="password-confirmation"
           aria-required="true"
           value={values.passwordConfirmation}
           onChange={handleChange}
           onBlur={handleBlur}
          />
          {dirtied.passwordConfirmation && errors.passwordConfirmation &&
            <FormError aria-describedby="password-confirmation">
              {errors.passwordConfirmation}
            </FormError>}
        </li>
      </FieldList>
      <SubmitButton type="submit" disabled={disabled} loading={submitting}>
        Sign up
      </SubmitButton>
    </StyledSignup>
  );
};

Signup.propTypes = {
  /**
   * Creates an object generating a function that returns a promise capable of
   * dispatching a series of async actions.
   */
  createAsyncFunction: PropTypes.func.isRequired,
};

export default Signup;
