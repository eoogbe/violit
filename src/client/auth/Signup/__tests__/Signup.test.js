/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import * as ReactRouterDom from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import * as USER from '__fixtures__/userFixture';
import { Button } from 'client/core';
import { FormError, Input } from 'client/forms';
import mockFetch from 'testUtils/mockFetch';
import { updateWrapper, wait } from 'testUtils/wait';
import actions from '../../actions';
import { NAME } from '../../constants';
import Signup from '../Signup';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(() => ({
    replace: () => {},
  })),
}));

const mockStore = configureStore();

const expectState = (initialState = {}) => {
  const store = mockStore({
    [NAME]: initialState,
  });

  let renderer;
  TestRenderer.act(() => {
    renderer = TestRenderer.create(
      <Provider store={store}>
        <ReactRouterDom.MemoryRouter initialEntries={['/signup']}>
          <Signup createAsyncFunction={() => {}} />
        </ReactRouterDom.MemoryRouter>
      </Provider>
    );
  });

  expect(renderer.toJSON()).toMatchSnapshot();
};

const mountSignup = (options) => {
  const { asyncFunction, initialEntries } = options || {};
  const store = mockStore();
  const createAsyncFunction = () => ({ asyncFunction });
  const wrapper = mount(
    <Provider store={store}>
      <ReactRouterDom.MemoryRouter initialEntries={initialEntries}>
        <Signup createAsyncFunction={createAsyncFunction} />
      </ReactRouterDom.MemoryRouter>
    </Provider>
  );
  return { store, wrapper };
};

const changeUsername = (wrapper, value) => {
  wrapper.find(Input).first().simulate('change', {
    target: {
      name: 'username',
      value,
    },
  });
};

const blurUsername = (wrapper, value) => {
  wrapper.find(Input).first().simulate('blur', {
    target: {
      name: 'username',
      value,
    },
  });
};

const changePassword = (wrapper, value) => {
  wrapper.find(Input).at(1).simulate('change', {
    target: {
      name: 'password',
      value,
    },
  });
};

const blurPassword = (wrapper, value) => {
  wrapper.find(Input).at(1).simulate('blur', {
    target: {
      name: 'password',
      value,
    },
  });
};

const changePasswordConfirmation = (wrapper, value) => {
  wrapper.find(Input).at(2).simulate('change', {
    target: {
      name: 'password-confirmation',
      value,
    },
  });
};

const blurPasswordConfirmation = (wrapper, value) => {
  wrapper.find(Input).at(2).simulate('blur', {
    target: {
      name: 'password-confirmation',
      value,
    },
  });
};

const simulateSubmit = async (wrapper, waitFn = wait) => {
  const onSubmit = wrapper.find(Signup).children().prop('onSubmit');
  return act(async () => {
    await onSubmit({
      preventDefault: () => {},
    });
    await waitFn();
  });
};

const getFieldErrorWrapper = (wrapper, index) =>
  wrapper
    .find('FieldList')
    .children()
    .children()
    .children()
    .at(index)
    .find(FormError);

it('renders without crashing', () => {
  expectState();
});

it('sets the document title', () => {
  const store = mockStore();
  TestRenderer.act(() => {
    TestRenderer.create(
      <Provider store={store}>
        <ReactRouterDom.MemoryRouter initialEntries={['/signup']}>
          <Signup createAsyncFunction={() => {}} />
        </ReactRouterDom.MemoryRouter>
      </Provider>
    );
  });
  expect(document.title).toContain('Signup');
});

describe('when submitting', () => {
  it('runs async function', async () => {
    mockFetch([
      USER.doesNotHaveUserFetch,
    ]);

    const asyncFunction = jest.fn();

    const { wrapper } = mountSignup({ asyncFunction });
    expect(wrapper.find(Button).prop('disabled')).toBe(true);

    changeUsername(wrapper, USER.username);
    blurUsername(wrapper, USER.username);

    changePassword(wrapper, USER.password);
    blurPassword(wrapper, USER.password);

    changePasswordConfirmation(wrapper, USER.password);
    blurPasswordConfirmation(wrapper, USER.password);

    await simulateSubmit(wrapper);

    expect(asyncFunction).toHaveBeenCalledWith({
      user: {
        username: USER.username,
        password: USER.password,
      },
    });
  });
  it('when previous location saved redirects to the previous location', async () => {
    mockFetch([
      USER.doesNotHaveUserFetch,
    ]);

    const historyReplace = jest.fn();
    ReactRouterDom.useHistory.mockReturnValue({
      replace: historyReplace,
    });

    const { wrapper } = mountSignup({
      asyncFunction: () => Promise.resolve(),
      initialEntries: [
        {
          pathname: '/signup',
          state: {
            from: '/prev',
          },
        },
      ],
    });

    changeUsername(wrapper, USER.username);
    blurUsername(wrapper, USER.username);

    changePassword(wrapper, USER.password);
    blurPassword(wrapper, USER.password);

    changePasswordConfirmation(wrapper, USER.password);
    blurPasswordConfirmation(wrapper, USER.password);

    await simulateSubmit(wrapper);

    expect(historyReplace).toHaveBeenCalledWith('/prev');
  });
  it('when previous location not saved redirects to homepage', async () => {
    mockFetch([
      USER.doesNotHaveUserFetch,
    ]);

    const historyReplace = jest.fn();
    ReactRouterDom.useHistory.mockReturnValue({
      replace: historyReplace,
    });

    const { wrapper } = mountSignup({
      asyncFunction: () => Promise.resolve(),
      initialEntries: ['/signup'],
    });

    changeUsername(wrapper, USER.username);
    blurUsername(wrapper, USER.username);

    changePassword(wrapper, USER.password);
    blurPassword(wrapper, USER.password);

    changePasswordConfirmation(wrapper, USER.password);
    blurPasswordConfirmation(wrapper, USER.password);

    await simulateSubmit(wrapper);

    expect(historyReplace).toHaveBeenCalledWith('/');
  });
  it('with submit error sets form error', async () => {
    mockFetch([
      USER.doesNotHaveUserFetch,
    ]);

    const error = 'invalid';
    const { wrapper } = mountSignup({
      asyncFunction: () => Promise.reject({ error }),
    });

    expect(wrapper.find(FormError).first()).toHaveLength(0);

    changeUsername(wrapper, USER.username);
    blurUsername(wrapper, USER.username);

    changePassword(wrapper, USER.password);
    blurPassword(wrapper, USER.password);

    changePasswordConfirmation(wrapper, USER.password);
    blurPasswordConfirmation(wrapper, USER.password);

    await simulateSubmit(wrapper, () => updateWrapper(wrapper));

    expect(wrapper.find(FormError).first().text()).toBe(error);
    expect(wrapper.find(Button).prop('disabled')).toBe(true);
  });
});

it('with invalid username sets username error', async () => {
  const { wrapper } = mountSignup();

  let errorWrapper = getFieldErrorWrapper(wrapper, 0);
  expect(errorWrapper).toHaveLength(0);

  changeUsername(wrapper, '');
  blurUsername(wrapper, '');

  await updateWrapper(wrapper);

  errorWrapper = getFieldErrorWrapper(wrapper, 0);
  expect(errorWrapper.text()).toBe('Required.');

  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});

it('with invalid password sets password error', async () => {
  const { wrapper } = mountSignup();

  let errorWrapper = getFieldErrorWrapper(wrapper, 1);
  expect(errorWrapper).toHaveLength(0);

  changePassword(wrapper, '');
  blurPassword(wrapper, '');

  await updateWrapper(wrapper);

  errorWrapper = getFieldErrorWrapper(wrapper, 1);
  expect(errorWrapper.text()).toBe('Required.');

  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});

it('with invalid password confirmation sets password confirmation error', async () => {
  const { wrapper } = mountSignup();

  let errorWrapper = getFieldErrorWrapper(wrapper, 2);
  expect(errorWrapper).toHaveLength(0);

  changePasswordConfirmation(wrapper, '');
  blurPasswordConfirmation(wrapper, '');

  await updateWrapper(wrapper);

  errorWrapper = getFieldErrorWrapper(wrapper, 2);
  expect(errorWrapper.text()).toBe('Required.');

  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});
