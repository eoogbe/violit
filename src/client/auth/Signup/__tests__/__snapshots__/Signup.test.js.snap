// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<form
  className="sc-dnqmqq bwQyKk"
  onSubmit={[Function]}
>
  <h2
    className="sc-iwsKbI eMpqKh"
  >
    Signup
  </h2>
  <ul
    className="sc-gZMcBi dOixHG"
  >
    <li>
      <label
        htmlFor="username"
      >
        Username
      </label>
      <input
        aria-required="true"
        autoFocus={true}
        className="sc-htoDjs kEmxlR"
        id="username"
        name="username"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
        value=""
      />
    </li>
    <li>
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        aria-required="true"
        className="sc-htoDjs kEmxlR"
        id="password"
        name="password"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
        value=""
      />
    </li>
    <li>
      <label
        htmlFor="password-confirmation"
      >
        Password confirmation
      </label>
      <input
        aria-required="true"
        className="sc-htoDjs kEmxlR"
        id="password-confirmation"
        name="password-confirmation"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
        value=""
      />
    </li>
  </ul>
  <button
    className="sc-ifAKCX hbBAat sc-gqjmRU kZCXhP"
    disabled={true}
    type="submit"
  >
    Sign up
  </button>
</form>
`;
